File: CODE/CPP/arc_vector_3.cpp
// -- IMPORTS

#define _USE_MATH_DEFINES
#include <cmath>
#include "arc_vector_3.hpp"

// -- OPERATIONS

void ArcVector3::setQuadraticArcPositionVector(
    const ArcVector3 & firstPositionVector,
    const ArcVector3 & firstForwardVector,
    const ArcVector3 & secondPositionVector,
    const ArcVector3 & secondBackwardVector,
    double interpolationFactor
    )
{
    double interpolationFactorComplement = 1.0 - interpolationFactor;

    double firstQuadraticPositionX = firstPositionVector.x + firstForwardVector.x * interpolationFactor;
    double firstQuadraticPositionY = firstPositionVector.y + firstForwardVector.y * interpolationFactor;
    double firstQuadraticPositionZ = firstPositionVector.z + firstForwardVector.z * interpolationFactor;
    double secondQuadraticPositionX = secondPositionVector.x + secondBackwardVector.x * interpolationFactorComplement;
    double secondQuadraticPositionY = secondPositionVector.y + secondBackwardVector.y * interpolationFactorComplement;
    double secondQuadraticPositionZ = secondPositionVector.z + secondBackwardVector.z * interpolationFactorComplement;

    double secondQuadraticPositionFactor = ( 1.0 - cos( M_PI * interpolationFactor ) ) * 0.5;

    x = firstQuadraticPositionX + ( secondQuadraticPositionX - firstQuadraticPositionX ) * secondQuadraticPositionFactor;
    y = firstQuadraticPositionY + ( secondQuadraticPositionY - firstQuadraticPositionY ) * secondQuadraticPositionFactor;
    z = firstQuadraticPositionZ + ( secondQuadraticPositionZ - firstQuadraticPositionZ ) * secondQuadraticPositionFactor;
}

// ~~

void ArcVector3::setSphericalArcPositionVector(
    const ArcVector3 & firstPositionVector,
    const ArcVector3 & firstForwardVector,
    const ArcVector3 & secondPositionVector,
    const ArcVector3 & secondBackwardVector,
    double interpolationFactor
    )
{
    double firstForwardPositionVectorX = firstPositionVector.x + firstForwardVector.x;
    double firstForwardPositionVectorY = firstPositionVector.y + firstForwardVector.y;
    double firstForwardPositionVectorZ = firstPositionVector.z + firstForwardVector.z;

    double secondBackwardPositionVectorX = secondPositionVector.x + secondBackwardVector.x;
    double secondBackwardPositionVectorY = secondPositionVector.y + secondBackwardVector.y;
    double secondBackwardPositionVectorZ = secondPositionVector.z + secondBackwardVector.z;

    double firstResidualVectorX = secondPositionVector.x - firstForwardPositionVectorX;
    double firstResidualVectorY = secondPositionVector.y - firstForwardPositionVectorY;
    double firstResidualVectorZ = secondPositionVector.z - firstForwardPositionVectorZ;
    double secondResidualVectorX = secondBackwardPositionVectorX - firstPositionVector.x;
    double secondResidualVectorY = secondBackwardPositionVectorY - firstPositionVector.y;
    double secondResidualVectorZ = secondBackwardPositionVectorZ - firstPositionVector.z;

    double firstOriginPositionVectorX = firstPositionVector.x + firstResidualVectorX;
    double firstOriginPositionVectorY = firstPositionVector.y + firstResidualVectorY;
    double firstOriginPositionVectorZ = firstPositionVector.z + firstResidualVectorZ;
    double secondOriginPositionVectorX = secondPositionVector.x - secondResidualVectorX;
    double secondOriginPositionVectorY = secondPositionVector.y - secondResidualVectorY;
    double secondOriginPositionVectorZ = secondPositionVector.z - secondResidualVectorZ;

    double interpolationFactorAngle = ( 1.0 - interpolationFactor ) * ( M_PI * 0.5 );
    double interpolationFactorCosinus = cos( interpolationFactorAngle );
    double interpolationFactorSinus = sin( interpolationFactorAngle );

    double firstSphericalPositionX = firstOriginPositionVectorX + firstForwardVector.x * interpolationFactorCosinus - firstResidualVectorX * interpolationFactorSinus;
    double firstSphericalPositionY = firstOriginPositionVectorY + firstForwardVector.y * interpolationFactorCosinus - firstResidualVectorY * interpolationFactorSinus;
    double firstSphericalPositionZ = firstOriginPositionVectorZ + firstForwardVector.z * interpolationFactorCosinus - firstResidualVectorZ * interpolationFactorSinus;
    double secondSphericalPositionX = secondOriginPositionVectorX + secondResidualVectorX * interpolationFactorCosinus + secondBackwardVector.x * interpolationFactorSinus;
    double secondSphericalPositionY = secondOriginPositionVectorY + secondResidualVectorY * interpolationFactorCosinus + secondBackwardVector.y * interpolationFactorSinus;
    double secondSphericalPositionZ = secondOriginPositionVectorZ + secondResidualVectorZ * interpolationFactorCosinus + secondBackwardVector.z * interpolationFactorSinus;

    double secondSphericalPositionFactor = 1.0 - interpolationFactorSinus;

    x = firstSphericalPositionX + ( secondSphericalPositionX - firstSphericalPositionX ) * secondSphericalPositionFactor;
    y = firstSphericalPositionY + ( secondSphericalPositionY - firstSphericalPositionY ) * secondSphericalPositionFactor;
    z = firstSphericalPositionZ + ( secondSphericalPositionZ - firstSphericalPositionZ ) * secondSphericalPositionFactor;
}

File: CODE/RUST/src/arc_vector_3.rs
// -- TYPES

#[derive( Debug, Clone, Copy )]
pub struct Vector3
{
    pub x: f64,
    pub y: f64,
    pub z: f64
}

// -- IMPLEMENTATIONS

impl Vector3
{
    // -- CONSTRUCTORS
    
    pub fn new(
        x: f64,
        y: f64,
        z: f64
        ) -> Self
    {
        Self { x, y, z }
    }

    // -- OPERATIONS

    pub fn set_quadratic_arc_position_vector(
        &mut self,
        first_position_vector: &Vector3,
        first_forward_vector: &Vector3,
        second_position_vector: &Vector3,
        second_backward_vector: &Vector3,
        interpolation_factor: f64
        )
    {
        let interpolation_factor_complement = 1.0 - interpolation_factor;

        let first_quadratic_position_x = first_position_vector.x + first_forward_vector.x * interpolation_factor;
        let first_quadratic_position_y = first_position_vector.y + first_forward_vector.y * interpolation_factor;
        let first_quadratic_position_z = first_position_vector.z + first_forward_vector.z * interpolation_factor;
        let second_quadratic_position_x = second_position_vector.x + second_backward_vector.x * interpolation_factor_complement;
        let second_quadratic_position_y = second_position_vector.y + second_backward_vector.y * interpolation_factor_complement;
        let second_quadratic_position_z = second_position_vector.z + second_backward_vector.z * interpolation_factor_complement;

        let second_quadratic_position_factor = ( 1.0 - ( std::f64::consts::PI * interpolation_factor ).cos() ) * 0.5;

        self.x = first_quadratic_position_x + ( second_quadratic_position_x - first_quadratic_position_x ) * second_quadratic_position_factor;
        self.y = first_quadratic_position_y + ( second_quadratic_position_y - first_quadratic_position_y ) * second_quadratic_position_factor;
        self.z = first_quadratic_position_z + ( second_quadratic_position_z - first_quadratic_position_z ) * second_quadratic_position_factor;
    }

    // ~~

    pub fn set_spherical_arc_position_vector(
        &mut self,
        first_position_vector: &Vector3,
        first_forward_vector: &Vector3,
        second_position_vector: &Vector3,
        second_backward_vector: &Vector3,
        interpolation_factor: f64
        )
    {
        let first_forward_position_vector_x = first_position_vector.x + first_forward_vector.x;
        let first_forward_position_vector_y = first_position_vector.y + first_forward_vector.y;
        let first_forward_position_vector_z = first_position_vector.z + first_forward_vector.z;

        let second_backward_position_vector_x = second_position_vector.x + second_backward_vector.x;
        let second_backward_position_vector_y = second_position_vector.y + second_backward_vector.y;
        let second_backward_position_vector_z = second_position_vector.z + second_backward_vector.z;

        let first_residual_vector_x = second_position_vector.x - first_forward_position_vector_x;
        let first_residual_vector_y = second_position_vector.y - first_forward_position_vector_y;
        let first_residual_vector_z = second_position_vector.z - first_forward_position_vector_z;
        let second_residual_vector_x = second_backward_position_vector_x - first_position_vector.x;
        let second_residual_vector_y = second_backward_position_vector_y - first_position_vector.y;
        let second_residual_vector_z = second_backward_position_vector_z - first_position_vector.z;

        let first_origin_position_vector_x = first_position_vector.x + first_residual_vector_x;
        let first_origin_position_vector_y = first_position_vector.y + first_residual_vector_y;
        let first_origin_position_vector_z = first_position_vector.z + first_residual_vector_z;
        let second_origin_position_vector_x = second_position_vector.x - second_residual_vector_x;
        let second_origin_position_vector_y = second_position_vector.y - second_residual_vector_y;
        let second_origin_position_vector_z = second_position_vector.z - second_residual_vector_z;

        let interpolation_factor_angle = ( 1.0 - interpolation_factor ) * ( std::f64::consts::PI * 0.5 );
        let interpolation_factor_cosinus = interpolation_factor_angle.cos();
        let interpolation_factor_sinus = interpolation_factor_angle.sin();

        let first_spherical_position_x = first_origin_position_vector_x + first_forward_vector.x * interpolation_factor_cosinus - first_residual_vector_x * interpolation_factor_sinus;
        let first_spherical_position_y = first_origin_position_vector_y + first_forward_vector.y * interpolation_factor_cosinus - first_residual_vector_y * interpolation_factor_sinus;
        let first_spherical_position_z = first_origin_position_vector_z + first_forward_vector.z * interpolation_factor_cosinus - first_residual_vector_z * interpolation_factor_sinus;
        let second_spherical_position_x = second_origin_position_vector_x + second_residual_vector_x * interpolation_factor_cosinus + second_backward_vector.x * interpolation_factor_sinus;
        let second_spherical_position_y = second_origin_position_vector_y + second_residual_vector_y * interpolation_factor_cosinus + second_backward_vector.y * interpolation_factor_sinus;
        let second_spherical_position_z = second_origin_position_vector_z + second_residual_vector_z * interpolation_factor_cosinus + second_backward_vector.z * interpolation_factor_sinus;

        let second_spherical_position_factor = 1.0 - interpolation_factor_sinus;

        self.x = first_spherical_position_x + ( second_spherical_position_x - first_spherical_position_x ) * second_spherical_position_factor;
        self.y = first_spherical_position_y + ( second_spherical_position_y - first_spherical_position_y ) * second_spherical_position_factor;
        self.z = first_spherical_position_z + ( second_spherical_position_z - first_spherical_position_z ) * second_spherical_position_factor;
    }
}
